### ISSUE TITLE
Enhancement: [no-unnecessary-condition] allow unnecessary `switch`/`case`

### ISSUE THREAD

**Issue by lensbart** [typescript-eslint/typescript-eslint#11678]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [x] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [x] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/no-unnecessary-condition

### Description

A prime use case for `switch`/`case` is to check exhaustiveness, so that you get warned when adding or removing an enum member later on. This exhaustiveness guarantee cannot be offered by successive `if`/`else` blocks.

When building an application, it can be useful to model something as an enum even though it only has one enum member for the time being. That way, when the app evolves you’ll be able to get a lint warning in all places that have unhandled enum members (via non-exhaustive `switch`/`case`).

In my case, I’m using React Router with the following structure for route actions:

```ts
import * as z from "zod/mini"

import { parseFormData } from "#src/utils/parseFormData.server.ts"
import type { Route } from "./+types/route.ts"

const actionSchema = z.discriminatedUnion("intent", [
	z.object({
		intent: z.literal("something"),
	}),
])

export const action = async ({ request }: Route.ActionArgs) => {
	const result = await parseFormData(request, actionSchema)
	if (!result.success) {
		return result
	}
	const { data } = result
	switch (data.intent) {
		// eslint-disable-next-line @typescript-eslint/no-unnecessary-condition -- placeholder structure
		case "something":
			break
	}
}
```

This is the general structure for every `action` route, to be extended later if the route is to handle more cases. Other examples exist (with the single-member enum being a database type, for example).

It would be handy if the `// eslint-disable-next-line @typescript-eslint/no-unnecessary-condition -- placeholder structure` would not be necessary in the example above.

Thanks for your consideration!

### Fail

```TypeScript
var replace = 'me';
```

### Pass

```TypeScript
const replace = 'me';
```

**Comment by bradzacher** [comment]
Duplicate of https://github.com/typescript-eslint/typescript-eslint/issues/10283

Our stance is that completely disabling the feature for an edge-case usecase like this is not a good enough motivation for us to build something. A disable comment works well and also will automatically flag itself for removal when you add more cases (as eslint reports disable comments that disable nothing).

My personal 2c is that if you're creating so many single-case switch statements in your codebase that this is an actual concern for you -- that in of itself is a bit of a code smell that perhaps your boilerplate is more verbose than actually required, and instead a simple type-only check like suggested in the above issue would be a better way to structure things until more cases are actually required.

**Comment by lensbart** [comment]
For what it’s worth, I implemented this as a custom ESLint rule as follows:

```ts
import assert from "node:assert/strict"

import rules from "@typescript-eslint/eslint-plugin/use-at-your-own-risk/rules"
import type { Rule } from "eslint"

const rule = rules["no-unnecessary-condition"]

assert(rule != null)

/**
 * A modified version of `@typescript-eslint/no-unnecessary-condition` that
 * ignores `switch`/`case` blocks with one sole enum member.
 *
 * See https://github.com/typescript-eslint/typescript-eslint/issues/11678
 */
export const noUnnecessaryCondition: Rule.RuleModule = {
	create(context) {
		const originalReport = context.report.bind(context)

		type Violation = Parameters<typeof originalReport>[0]

		const wrappedContext = Object.assign(
			Object.create(Object.getPrototypeOf(context)),
			context,
			{
				report(violation: Violation) {
					if (
						"node" in violation &&
						typeof violation.node === "object" &&
						"parent" in violation.node &&
						violation.node.parent != null &&
						typeof violation.node.parent === "object" &&
						"type" in violation.node.parent &&
						violation.node.parent.type === "SwitchCase"
					) {
						return
					}
					return originalReport(violation)
				},
			},
		)

		return rule.create(wrappedContext) as Rule.RuleListener
	},
	meta: rule.meta,
}
```

The code was LLM-generated and is neither clean nor type-safe, but seems functional. Sharing it here in case it could be of interest.

If you include this custom ESLint rule, `@typescript-eslint/no-unnecessary-condition` can be disabled and all errors should still be flagged as before, except for the `switch` statement.

@bradzacher I can’t comment in https://github.com/typescript-eslint/typescript-eslint/issues/11448 and https://github.com/typescript-eslint/typescript-eslint/issues/10283, but I think this might be useful to the people who participated in that thread as well. Would you kindly share it over there?

Thanks.
